esp-hal/src/rmt.rs

```diff
@@ -250,59 +250,235 @@ pub enum Error {
     MemoryBlockNotAvailable,
 }
 
-///  Convenience trait to work with pulse codes.
-pub trait PulseCode: crate::private::Sealed {
-    /// Create a new instance
-    fn new(level1: Level, length1: u16, level2: Level, length2: u16) -> Self;
+///  Convenience newtype to work with pulse codes.
+#[derive(Clone, Copy, Default, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct PulseCode(u32);
 
-    /// Create a new empty instance
-    fn empty() -> Self;
+// Use these constants such that the code below is exactly the same for both halves of the
+// PulseCode and it is easier to verify that they're not mixed up.
+const LEVEL1_SHIFT: usize = 0;
+const LEVEL2_SHIFT: usize = 16;
 
-    /// Set all levels and lengths to 0
-    fn reset(&mut self);
+impl PulseCode {
+    /// Create a new instance.
+    ///
+    /// SAFETY:
+    /// - `length1` and `length2` must be 15-bit wide, i.e. their MSB must be cleared.
+    #[inline]
+    const unsafe fn new_unchecked(
+        level1: Level,
+        length1: u16,
+        level2: Level,
+        length2: u16,
+    ) -> Self {
+        let mut code = 0;
 
-    /// Logical output level in the first pulse code interval
-    fn level1(&self) -> Level;
+        if matches!(level1, Level::High) {
+            code |= 1 << (LEVEL1_SHIFT + 15);
+        }
+        code |= (length1 as u32) << LEVEL1_SHIFT;
 
-    /// Length of the first pulse code interval (in clock cycles)
-    fn length1(&self) -> u16;
+        if matches!(level2, Level::High) {
+            code |= 1 << (LEVEL2_SHIFT + 15);
+        }
+        code |= (length2 as u32) << LEVEL2_SHIFT;
+
+        Self(code)
+    }
+
+    /// Create a new instance.
+    ///
+    /// If `length1` or `length2` exceed the maximum representable range, they
+    /// will be clamped to `0x7FFF`.
+    #[inline]
+    pub const fn new(level1: Level, length1: u16, level2: Level, length2: u16) -> Self {
+        // Can't use lengthX.min(0x7FFF) since it is not const
+        let length1 = if length1 >= 0x8000 { 0x7FFF } else { length1 };
```

Comment:
```suggestion
        let length1 = length1.min(0x7FFF);
```

Reply:
Unfortunately, that's not `const` due to the lack of support for `const fn` in the `Ord` trait: https://internals.rust-lang.org/t/int-min-max-in-const-fn/23232

Reply:
I keep not noticing `const` today...